src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/HeaderBuilder.java

```diff
@@ -60,20 +61,24 @@ public void addFunctionalInterface(String name, MethodType mtype,  FunctionDescr
         indent();
     }
 
-    public void addStaticFunctionWrapper(String javaName, String nativeName, MethodType mtype, FunctionDescriptor desc, boolean varargs, List<String> paramNames) {
+    public void addStaticFunctionWrapper(String javaName, String nativeName, MethodType sigtype, MethodType mhtype, FunctionDescriptor desc, boolean varargs, List<String> paramNames) {
         incrAlign();
         indent();
-        sb.append(PUB_MODS + mtype.returnType().getName() + " " + javaName + " (");
+        sb.append(PUB_MODS + sigtype.returnType().getName() + " " + javaName + " (");
         String delim = "";
-        List<String> pNames = new ArrayList<>();
+        List<String> pExprs = new ArrayList<>();
         final int numParams = paramNames.size();
         for (int i = 0 ; i < numParams; i++) {
             String pName = paramNames.get(i);
             if (pName.isEmpty()) {
                 pName = "x" + i;
             }
-            pNames.add(pName);
-            sb.append(delim + mtype.parameterType(i).getName() + " " + pName);
+            if (sigtype.parameterType(i).equals(Addressable.class)) {
```

Comment:
Here we assume the method handle have MemoryAddress type, while this is true in current implementation, it's possible that mhtype actually have Addressable if caller chose to do so otherwise.

Reply:
No, that's not possible. The ABI does not support Addressable as a carrier type for parameter/returns. In other words, Addressable is a fiction that's only valid in jextract - it is not really taken into account by the guts of the ABI support.

Reply:
It is true at ABI level, the method handle can only handle MemoryAddress. But there is nothing preventing something like ConstantHelper to adapt the ABI-level MH to take Addressable as you do here.

What I am suggesting is simply saying you don't need add sigtype, simply take the incoming method type and adapt MemoryAddress into Addressable.address().

Reply:
> What I am suggesting is simply saying you don't need add sigtype, simply take the incoming method type and adapt MemoryAddress into Addressable.address().

Not sure I get it. I think this same builder code is reused both for downcalls and upcalls; but we don't want to do the transformation for upcalls. So at the very least we'd need to pass a boolean to turn the transformation on/off. If we go down this path, then I guess TypeTranslator could be reverted back to what it was - is that what you are suggesting?